heuristic/heuristic_calculations.py

```diff
@@ -1,322 +1,211 @@
+                delta[c,t] = max(0, model.demand["ideal"][c,t] - sum(y[c,e,t] for e in model.employee_with_competencies[c]))
+                # if(delta[c,t] - abs(model.delta["plus"][c,t].x - model.delta["minus"][c,t].x)) != 0:
+                #     print(delta[c,t] - abs(model.delta["plus"][c,t].x - model.delta["minus"][c,t].x))
+                #     print("Different Delta")
     return delta
 
-
-def calculate_negative_deviation_from_contracted_hours(data, y):
-
-    time_periods_in_week = data["time"]["periods"][2]
+def calculate_negative_deviation_from_contracted_hours(model, y):
     delta_negative_contracted_hours = {}
-
-    for e in data["staff"]["employees"]:
-        for j in data["time"]["weeks"]:
-            delta_negative_contracted_hours[e, j] = data["staff"]["employee_contracted_hours"][
-                e
-            ] - sum(
-                data["time"]["step"] * y[c, e, t]
-                for t in time_periods_in_week[j]
-                for c in data["competencies"]
-            )
+    for e in model.employees:
+        for j in model.weeks:
+            delta_negative_contracted_hours[e,j] = (model.contracted_hours[e]
+            - sum(model.time_step * y[c,e,t] 
+            for c in model.competencies
+            for t in model.time_periods_in_week[c, j]))
     return delta_negative_contracted_hours
 
-
-def calculate_partial_weekends(data, x):
-
+def calculate_partial_weekends(model, x):
     partial_weekend = {}
-
-    for i in data["time"]["saturdays"]:
-        for e in data["staff"]["employees"]:
-
-            partial_weekend[e, i] = abs(
-                sum(x[e, t, v] for t, v in data["shifts"]["shifts_per_day"][i])
-                - sum(x[e, t, v] for t, v in data["shifts"]["shifts_per_day"][i + 1])
-            )
+    partial_weekend_shifts = []
+    for i in model.saturdays:
+        for e in model.employees:
+            #if(abs(sum(model.x[e,t,v] for t,v in model.shifts_at_day[i]) - sum(x[e,t,v] for t,v in model.shifts_at_day[i+1])) != 0):
+             #   partial_weekend_shifts.extend([(e,t,v) for t,v in model.shifts_at_day[i] if x[e,t,v] == 1])
+              #  partial_weekend_shifts.extend([(e,t,v) for t,v in model.shifts_at_day[i+1] if x[e,t,v] == 1])
+
+            partial_weekend[e,i] =  abs((sum(x[e,t,v] 
+                                    for t,v in model.shifts_at_day[i]) 
+                                    - sum(x[e,t,v] 
+                                    for t,v in model.shifts_at_day[i+1])))
+            # if(partial_weekend[e,i] != abs(model.ro["sat"][e,i].x - model.ro["sun"][e,(i+1)].x)):
+            #     print(str(partial_weekend) + "," + str(abs(model.ro["sat"][e,i].x - model.ro["sun"][e,(i+1)].x)))
+            #     print(i)
+            #     print("Different partial weekends")
     return partial_weekend
 
-
-def calculate_isolated_working_days(data, x):
-
-    shifts_per_day = data["shifts"]["shifts_per_day"]
+def calculate_isolated_working_days(model, x):
```

Comment:
Just a thought: we could consider calculating `isolated_working_days` and `isolated_off_days` in the same loop.

Reply:
Yes that is a good idea.